arangod/RocksDBEngine/RocksDBChecksumEnv.cpp

```diff
@@ -0,0 +1,264 @@
+/// limitations under the License.
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Julia Puget
+////////////////////////////////////////////////////////////////////////////////
+
+#include "RocksDBChecksumEnv.h"
+
+#include "Basics/Exceptions.h"
+#include "Basics/files.h"
+#include "Basics/FileUtils.h"
+#include "Basics/MutexLocker.h"
+#include "Logger/LogMacros.h"
+
+namespace arangodb::checksum {
+
+ChecksumCalculator::ChecksumCalculator()
+    :
+#if OPENSSL_VERSION_NUMBER >= 0x10100000L
+      _context(EVP_MD_CTX_new()) {
+#else
+      _context(EVP_MD_CTX_create()) {
+#endif
+  if (_context == nullptr) {
+    THROW_ARANGO_EXCEPTION(TRI_ERROR_OUT_OF_MEMORY);
+  }
+  if (EVP_DigestInit_ex(_context, EVP_sha256(), nullptr) == 0) {
+    THROW_ARANGO_EXCEPTION_MESSAGE(TRI_ERROR_INTERNAL,
+                                   "unable to initialize SHA256 processor");
+  }
+}
+
+void ChecksumCalculator::computeFinalChecksum() {
+  TRI_ASSERT(_context != nullptr);
+  unsigned char hash[EVP_MAX_MD_SIZE];
+  unsigned int lengthOfHash = 0;
+  if (EVP_DigestFinal_ex(_context, hash, &lengthOfHash) == 0) {
+    TRI_ASSERT(false);
+  }
+  _checksum = basics::StringUtils::encodeHex(
+      reinterpret_cast<char const*>(&hash[0]), lengthOfHash);
+}
+
+void ChecksumCalculator::updateIncrementalChecksum(char const* buffer,
+                                                   size_t n) {
+  TRI_ASSERT(_context != nullptr);
+  updateEVPWithContent(buffer, n);
+}
+
+ChecksumCalculator::~ChecksumCalculator() {
+#if OPENSSL_VERSION_NUMBER >= 0x10100000L
+  EVP_MD_CTX_free(_context);
+#else
+  EVP_MD_CTX_destroy(_context);
+#endif
+}
+
+void ChecksumCalculator::updateEVPWithContent(const char* buffer, size_t n) {
+  TRI_ASSERT(_context != nullptr);
+  if (EVP_DigestUpdate(_context, static_cast<void const*>(buffer), n) == 0) {
+    TRI_ASSERT(false);
+  }
+}
+
+bool ChecksumHelper::isFileNameSst(std::string const& fileName) {
+  return TRI_Basename(fileName).size() > 4 &&
+         (fileName.compare(fileName.size() - 4, 4, ".sst") == 0);
+}
+
+bool ChecksumHelper::writeShaFile(std::string const& fileName,
+                                  std::string const& checksum) {
+  TRI_ASSERT(isFileNameSst(fileName));
+
+  std::string shaFileName = buildShaFileNameFromSst(fileName, checksum);
+
+  LOG_TOPIC("80257", DEBUG, arangodb::Logger::ENGINES)
+      << "shaCalcFile: done " << fileName << " result: " << shaFileName;
+  auto res = TRI_WriteFile(shaFileName.c_str(), "", 0);
+  if (res == TRI_ERROR_NO_ERROR) {
+    {
+      MUTEX_LOCKER(mutexLock, _calculatedHashesMutex);
+      _fileNamesToHashes.try_emplace(TRI_Basename(fileName), checksum);
+    }
+    return true;
+  }
+
+  LOG_TOPIC("8f7ef", WARN, arangodb::Logger::ENGINES)
+      << "shaCalcFile: writing file failed with " << res << " for "
+      << shaFileName;
+  return false;
+}
+
+void ChecksumHelper::checkMissingShaFiles() {
```

Comment:
As far as I see this function is used in two places: 

(1) At startup, when the main db directory is checked. It also updates the member `_fileNamesToHashes` after a restart.
(2) It is called on a hotbackup when that is created.

I wonder whether this second call would confuse the global member `_fileNamesToHashes`?

Reply:
I think my comment is rubbish. I finally understood that a shaFilesManager object is created especially for this purpose in (2) and so all seems to be good.